tests/components/honeywell/test_init.py

```diff
@@ -29,3 +29,16 @@ async def test_setup_multiple_thermostats(
     await hass.async_block_till_done()
     assert config_entry.state is ConfigEntryState.LOADED
     assert hass.states.async_entity_ids_count() == 2
+
+
+@patch("homeassistant.components.honeywell.UPDATE_LOOP_SLEEP_TIME", 0)
+async def test_setup_multiple_thermostats_with_same_deviceid(
+    hass: HomeAssistant, config_entry: MockConfigEntry, location, device
+) -> None:
+    """Test that the config form is shown."""
+    location.devices_by_id[device.deviceid] = device
```

Comment:
I don't think this does anything because this map already contains device.deviceid? Or were you able to actually reproduce the log message with this?

I don't think there is anything necessarily wrong with this PR, i'm just worried it doesn't actually address the issue and we don't quite understand it yet.

Reply:
Do you need multiple locations instead for this to be able to happen?

Reply:
Unfortunately the problem here is that I do not know what the honeywell API is returning that causes this issue. The running theory was that it's returning the same device twice somehow. It could be under different locations, but in my case I only have a single location on my honeywell account. Perhaps it's duplicating the locations and devices? I've been running the changes for the past 4 days and had 1 `SomeComfort update failed` error, and the integration continued working after that.

You're correct though, if I take this test and run it against `dev` branch the test will still pass. So I need to figure out how I can make this test red on the `dev` branch before I finish this PR up.

Reply:
Agreed, making the test fail in the dev branch sounds like the right idea.

In order to get the same device added twice I think you're going to need to do something like this:

```
...
def test_example(
    ....
    device,
    client,
)...
    # set up a second location that refers to the original device id, this will cause the API to return two devices with the same id
    mock_location2 = create_autospec(somecomfort.Location, instance=True)
    mock_location2.locationid.return_value = "location2"
    mock_location2.devices_by_id = {device.deviceid: device}
    client.locations_by_id["location2"] = mock_location2
```

Reply:
The challenge here is that the entity registry will detect the duplicate device and just not add it. This will log an error but no exception is actually thrown or anything like that so it's difficult to assert on. Using the code you posted above (thanks by the way!) I can see the unique ID error if I assert on something that will always return a failed test. But the question remains: how do I assert on a logger entry?

Reply:
Check out `caplog`

Reply:
And there we have it, a red test. Thanks so much for helping me with this. Still developing my Python skills haha.

```
pytest test_init.py
Test session starts (platform: linux, Python 3.9.7, pytest 6.2.5, pytest-sugar 0.9.4)
rootdir: /workspaces/core, configfile: pyproject.toml
plugins: rerunfailures-10.2, mock-3.6.1, socket-0.4.1, respx-0.19.0, freezegun-0.4.2, timeout-2.0.1, sugar-0.9.4, xdist-2.4.0, forked-1.3.0, test-groups-1.0.3, requests-mock-1.9.2, cov-2.12.1, aiohttp-0.3.0, anyio-3.3.2
collecting ... 
 tests/components/honeywell/test_init.py ✓✓                                                                                                        67% ██████▋   

――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_example[pyloop] ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

hass = <homeassistant.core.HomeAssistant object at 0x7fa74548abe0>, caplog = <_pytest.logging.LogCaptureFixture object at 0x7fa74548aaf0>
config_entry = <tests.common.MockConfigEntry object at 0x7fa745e19e80>, device = <NonCallableMagicMock spec='Device' id='140356398463392'>
client = <NonCallableMagicMock name='SomeComfort()' spec='SomeComfort' id='140356398752768'>

    @patch("homeassistant.components.honeywell.UPDATE_LOOP_SLEEP_TIME", 0)
    async def test_example(
        hass: HomeAssistant, caplog, config_entry: MockConfigEntry, device, client
    ) -> None:
        """Test Honeywell TCC API returning duplicate device IDs."""
        mock_location2 = create_autospec(somecomfort.Location, instance=True)
        mock_location2.locationid.return_value = "location2"
        mock_location2.devices_by_id = {device.deviceid: device}
        client.locations_by_id["location2"] = mock_location2
        config_entry.add_to_hass(hass)
        await hass.config_entries.async_setup(config_entry.entry_id)
        await hass.async_block_till_done()
        assert config_entry.state is ConfigEntryState.LOADED
        assert hass.states.async_entity_ids_count() == 1
>       assert (
            "Platform honeywell does not generate unique IDs"
            not in caplog.text
        )
E       AssertionError: assert 'Platform ho...e unique IDs' not in 'INFO     ho...te.device1\n'
E         'Platform honeywell...generate unique IDs' is contained here:
E           rm.py:592 Platform honeywell does not generate unique IDs. ID 1234567 already exists - ignoring climate.device1

test_init.py:49: AssertionError
--------------------------------------------------------------------- Captured stderr setup ---------------------------------------------------------------------
DEBUG:asyncio:Using selector: EpollSelector
---------------------------------------------------------------------- Captured log setup -----------------------------------------------------------------------
DEBUG    asyncio:selector_events.py:59 Using selector: EpollSelector
--------------------------------------------------------------------- Captured stderr call ----------------------------------------------------------------------
INFO:homeassistant.loader:Loaded honeywell from homeassistant.components.honeywell
INFO:homeassistant.setup:Setting up honeywell
INFO:homeassistant.setup:Setup of domain honeywell took 0.0 seconds
DEBUG:homeassistant.core:Bus:Handling <Event component_loaded[L]: component=honeywell>
INFO:homeassistant.loader:Loaded climate from homeassistant.components.climate
INFO:homeassistant.setup:Setting up climate
DEBUG:homeassistant.core:Bus:Handling <Event service_registered[L]: domain=climate, service=turn_on>
DEBUG:homeassistant.core:Bus:Handling <Event service_registered[L]: domain=climate, service=turn_off>
DEBUG:homeassistant.core:Bus:Handling <Event service_registered[L]: domain=climate, service=set_hvac_mode>
DEBUG:homeassistant.core:Bus:Handling <Event service_registered[L]: domain=climate, service=set_preset_mode>
DEBUG:homeassistant.core:Bus:Handling <Event service_registered[L]: domain=climate, service=set_aux_heat>
DEBUG:homeassistant.core:Bus:Handling <Event service_registered[L]: domain=climate, service=set_temperature>
DEBUG:homeassistant.core:Bus:Handling <Event service_registered[L]: domain=climate, service=set_humidity>
DEBUG:homeassistant.core:Bus:Handling <Event service_registered[L]: domain=climate, service=set_fan_mode>
DEBUG:homeassistant.core:Bus:Handling <Event service_registered[L]: domain=climate, service=set_swing_mode>
INFO:homeassistant.setup:Setup of domain climate took 0.0 seconds
DEBUG:homeassistant.core:Bus:Handling <Event component_loaded[L]: component=climate>
INFO:homeassistant.components.climate:Setting up climate.honeywell
INFO:homeassistant.helpers.entity_registry:Registered new climate.honeywell entity: climate.device1
DEBUG:homeassistant.core:Bus:Handling <Event entity_registry_updated[L]: action=create, entity_id=climate.device1>
DEBUG:homeassistant.core:Bus:Handling <Event state_changed[L]: entity_id=climate.device1, old_state=None, new_state=<state climate.device1=off; hvac_modes=['off', 'heat_cool', 'cool', 'heat'], min_temp=None, max_temp=None, preset_modes=['none', 'away', 'Hold'], current_temperature=-6.7, temperature=None, target_temp_high=None, target_temp_low=None, current_humidity=<MagicMock name='mock.current_humidity' id='140356398663664'>, preset_mode=None, aux_heat=off, fan_action=running, permanent_hold=False, dr_phase=<MagicMock name='mock.raw_dr_data.get()' id='140356398911984'>, friendly_name=device1, supported_features=83 @ 2021-12-09T15:01:59.687082+00:00>>
ERROR:homeassistant.components.climate:Platform honeywell does not generate unique IDs. ID 1234567 already exists - ignoring climate.device1
----------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------
INFO     homeassistant.loader:loader.py:376 Loaded honeywell from homeassistant.components.honeywell
INFO     homeassistant.setup:setup.py:198 Setting up honeywell
INFO     homeassistant.setup:setup.py:246 Setup of domain honeywell took 0.0 seconds
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event component_loaded[L]: component=honeywell>
INFO     homeassistant.loader:loader.py:376 Loaded climate from homeassistant.components.climate
INFO     homeassistant.setup:setup.py:198 Setting up climate
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event service_registered[L]: domain=climate, service=turn_on>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event service_registered[L]: domain=climate, service=turn_off>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event service_registered[L]: domain=climate, service=set_hvac_mode>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event service_registered[L]: domain=climate, service=set_preset_mode>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event service_registered[L]: domain=climate, service=set_aux_heat>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event service_registered[L]: domain=climate, service=set_temperature>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event service_registered[L]: domain=climate, service=set_humidity>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event service_registered[L]: domain=climate, service=set_fan_mode>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event service_registered[L]: domain=climate, service=set_swing_mode>
INFO     homeassistant.setup:setup.py:246 Setup of domain climate took 0.0 seconds
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event component_loaded[L]: component=climate>
INFO     homeassistant.components.climate:entity_platform.py:235 Setting up climate.honeywell
INFO     homeassistant.helpers.entity_registry:entity_registry.py:397 Registered new climate.honeywell entity: climate.device1
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event entity_registry_updated[L]: action=create, entity_id=climate.device1>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event state_changed[L]: entity_id=climate.device1, old_state=None, new_state=<state climate.device1=off; hvac_modes=['off', 'heat_cool', 'cool', 'heat'], min_temp=None, max_temp=None, preset_modes=['none', 'away', 'Hold'], current_temperature=-6.7, temperature=None, target_temp_high=None, target_temp_low=None, current_humidity=<MagicMock name='mock.current_humidity' id='140356398663664'>, preset_mode=None, aux_heat=off, fan_action=running, permanent_hold=False, dr_phase=<MagicMock name='mock.raw_dr_data.get()' id='140356398911984'>, friendly_name=device1, supported_features=83 @ 2021-12-09T15:01:59.687082+00:00>>
ERROR    homeassistant.components.climate:entity_platform.py:592 Platform honeywell does not generate unique IDs. ID 1234567 already exists - ignoring climate.device1

 tests/components/honeywell/test_init.py ⨯                                                                                                        100% ██████████
==================================================================== short test summary info ====================================================================
FAILED test_init.py::test_example[pyloop] - AssertionError: assert 'Platform ho...e unique IDs' not in 'INFO     ho...te.device1\n'

Results (0.80s):
       2 passed
       1 failed
         - tests/components/honeywell/test_init.py:35 test_example[pyloop]
```

I'll get this test committed to the branch ASAP.